### DISCUSSION TITLE
Custom Color Map

### DISCUSSION THREAD

**Discussion by sammertime** [developmentseed/titiler#588] [Q&A]
Hi there,

First off - I love this package/project..Thank you! : )

I've reviewed the custom color map issues raised prior (i.e. below) and have successfully implemented the interval colormap into my tiles using titiler, as described here:

https://github.com/developmentseed/titiler/discussions/496

I've applied transparency to some of my intervals i.e. say for the following two intervals, the 295-296 range should be transparent:

[[294,295],[255,234,0,255]],[[295,296],[0,0,0,0]]
 
This is technically working, but transparency is returning as a black color in the tiles.  My goal is to have intervals with the color [0,0,0,0] in my custom colormap return as completely transparent from titiler.  

The query parameters I'm using with the tile endpoint are the following: 

bidx=1&unscale=false&resampling=nearest&return_mask=true&colormap=[[[...]]]

My question is, is this something that can be achieved with titiler and, if so, can it be done by tweaking my query parameters? Or, is this a more involved fix in the underlying COG creation process?

Thanks so much..
S

**Comment by vincentsarago** [comment]
sounds like https://github.com/developmentseed/titiler/discussions/525

I think the issue is that you are using `auto` formatting which sadly is decided before applying the colormap, if titiler don't see any mask value it will choose JPEG and then the [0,0,0,0] value will be black. If you force the image format to png it should be transparent. (I totally understand that it's not perfect, and we might try to fix this in rio-tiler at one point)

FYI: I'm moving this issue to discussion because without proper reproductible example I cannot understand if it's a `bug` or a `feature`.

**Reply by vincentsarago** [reply]
>  (I totally understand that it's not perfect, and we might try to fix this in rio-tiler at one point)

Turns out I was planning to fix this in next titiler version because I already made some change in rio-tiler 😅

**Reply by sammertime** [reply]
Ha! Well, thanks very much Vincent!  For both the prompt reply and the easy fix (my favorite kind!)

I forced the format to png and..voila!.. I can't believe I didn't try that after all my testing - I guess I'm no expert in image formats, but once you mentioned it I immediately remembered reading about differences between jpg and png..

In any case, thanks so much again and amazing work!

All the best,
S

**Reply by neelduttahere** [reply]
Hey, I would like to add to this discussion. I am trying to use a custom colormap for rendering interval-based output:

`[[[-4, 0.2],[127, 17, 224, 1]],[[0.2, 0.7],[242, 238, 189, 1]],[[0.7, 14],[227, 5, 5, 1]]]`

And below is my titiler request link
`${process.env.REACT_APP_COG_TILER}/cog/tiles/WebMercatorQuad/{z}/{x}/{y}?url=${output}&resampling_method=nearest&colormap=%5B%5B%5B-4%2C%200.2%5D%2C%5B127%2C%2017%2C%20224%2C%201%5D%5D%2C%5B%5B0.2%2C%200.7%5D%2C%5B242%2C%20238%2C%20189%2C%201%5D%5D%2C%5B%5B0.7%2C%2014%5D%2C%5B227%2C%205%2C%205%2C%201%5D%5D%5D%0A
                                &return_mask=false&nodata=-99`

Here I have pre-encode the colormap.

But here is an issue: I am getting a black color for where we have `nodata` or `nan` (see image for reference).


<img width="556" alt="Screenshot 2023-06-19 at 2 24 27 PM" src="https://github.com/developmentseed/titiler/assets/59992993/68f452d5-e1e0-4862-bdf9-a84527f9d1f4">

When I remove `nodata=-99` is get the result below:
<img width="333" alt="Screenshot 2023-06-19 at 2 27 39 PM" src="https://github.com/developmentseed/titiler/assets/59992993/6b025991-53ea-4070-9804-fbb82f7ad556">

Now when I change return_mask=true, any viewport image tile, that has `no-data` or `nan` in the raw tif file, for it returns a transparent tile, which looks very broken, like the below image:

<img width="729" alt="Screenshot 2023-06-19 at 2 32 13 PM" src="https://github.com/developmentseed/titiler/assets/59992993/a0003f0e-16a7-4d13-aa10-8fa7d55685d6">

Looking for suggestions on how to go about this issue. Thanks in advance!

**Reply by vincentsarago** [reply]
@neelduttahere which version of titiler are you using? 

as said in https://github.com/developmentseed/titiler/discussions/588#discussioncomment-4968039 if you force the output to be `png` you should see the black tiles. 

I'm pretty sure we fixed this issue in https://github.com/developmentseed/titiler/blob/main/CHANGES.md#0111a0-2023-03-01

**Reply by neelduttahere** [reply]
if I force the output to `png` this issue still exists. I am using version `0.11.7`

**Reply by vincentsarago** [reply]
@neelduttahere are you able to share your image?

**Reply by neelduttahere** [reply]
Sure. Here it is:
https://drive.google.com/file/d/1c0htdRFvfqVXAY3qT3Urfzep9SBzLGS2/view?usp=sharing

**Reply by vincentsarago** [reply]
@neelduttahere I can't really replicate. 

using 
```
http://127.0.0.1:8081/cog/preview?url=test-file.tif&colormap=%5B%5B%5B-4%2C+0.2%5D%2C+%5B127%2C+17%2C+224%2C+255%5D%5D%2C+%5B%5B0.2%2C+0.7%5D%2C+%5B242%2C+238%2C+189%2C+255%5D%5D%2C+%5B%5B0.7%2C+14%5D%2C+%5B227%2C+5%2C+5%2C+255%5D%5D%5D
```

I get

<img width="714" alt="Screenshot 2023-06-19 at 10 16 29 PM" src="https://github.com/developmentseed/titiler/assets/10407788/b3900fa2-af35-42f5-ac40-5a47f205bfa3">


```python
import urllib

# original cmap was 
# [[[-4, 0.2],[127, 17, 224, 1]],[[0.2, 0.7],[242, 238, 189, 1]],[[0.7, 14],[227, 5, 5, 1]]]
# but the colormap has to be in GDAL form and alpha value is either 0 or 255

cmap = [[[-4, 0.2],[127, 17, 224, 255]],[[0.2, 0.7],[242, 238, 189, 255]],[[0.7, 14],[227, 5, 5, 255]]]

urllib.parse.urlencode({"colormap": cmap})

>> "colormap=%5B%5B%5B-4%2C+0.2%5D%2C+%5B127%2C+17%2C+224%2C+255%5D%5D%2C+%5B%5B0.2%2C+0.7%5D%2C+%5B242%2C+238%2C+189%2C+255%5D%5D%2C+%5B%5B0.7%2C+14%5D%2C+%5B227%2C+5%2C+5%2C+255%5D%5D%5D"
```

**Reply by vincentsarago** [reply]
confirmed in the `/map` endpoint 

```
http://127.0.0.1:8081/cog/map?url=test-file.tif&colormap=%5B%5B%5B-4%2C+0.2%5D%2C+%5B127%2C+17%2C+224%2C+255%5D%5D%2C+%5B%5B0.2%2C+0.7%5D%2C+%5B242%2C+238%2C+189%2C+255%5D%5D%2C+%5B%5B0.7%2C+14%5D%2C+%5B227%2C+5%2C+5%2C+255%5D%5D%5D
```
<img width="837" alt="Screenshot 2023-06-19 at 10 22 07 PM" src="https://github.com/developmentseed/titiler/assets/10407788/08e4c12c-a1ca-4ef7-adc0-49f01ce60e56">

**Reply by vincentsarago** [reply]
oh and note that `return_mask` has to be set to `true` or the PNG won't have alpha band 

```
http://127.0.0.1:8081/cog/map?url=test-file.tif&colormap=%5B%5B%5B-4%2C+0.2%5D%2C+%5B127%2C+17%2C+224%2C+255%5D%5D%2C+%5B%5B0.2%2C+0.7%5D%2C+%5B242%2C+238%2C+189%2C+255%5D%5D%2C+%5B%5B0.7%2C+14%5D%2C+%5B227%2C+5%2C+5%2C+255%5D%5D%5D&return_mask=False
```
<img width="806" alt="Screenshot 2023-06-19 at 10 26 29 PM" src="https://github.com/developmentseed/titiler/assets/10407788/4be93ca8-25fd-43d1-a93b-be1be15c73ca">

**Reply by neelduttahere** [reply]
Hey so setting `return_mask=true` or not passing `return_mask` at all (since the default is true for it), gives a broken output as shown in the above responses. But for you it seems to work, I guess I need to look back into the dependencies installed, probably has to do with rio-tiler because last I remember we did have to downgrade `rio-tiler` for a specific issue. Will keep posted if anything arises

**Reply by neelduttahere** [reply]
@vincentsarago I notice that you are sending color information for each interval like this `[[-4, 0.2],[127, 17, 224, 255]]`, instead of this  `[[-4, 0.2],[127, 17, 224, 1]]`. I thought we send RGBA values against each interval, where the last number is for alpha (from `0.1 to 1`). Either way `[[-4, 0.2],[127, 17, 224, 255]]` works amazingly! Though curious why we have to send 255 instead decimal alpha.

Thanks a lot!

**Reply by vincentsarago** [reply]
> we have to send 255 instead decimal alpha.

That's a GDAL format, alpha value can be between 0 (transparent) and 255 (non-transparent).

https://cogeotiff.github.io/rio-tiler/colormap/
